src/PowerShellEditorServices.Protocol/MessageProtocol/Channel/NamedPipeServerListener.cs

```diff
@@ -89,47 +117,58 @@ public override void Start()
 
         public override void Stop()
         {
-            if (this.pipeServer != null)
+            if (this.inOutPipeServer != null)
             {
                 this.logger.Write(LogLevel.Verbose, "Named pipe server shutting down...");
 
-                this.pipeServer.Dispose();
+                this.inOutPipeServer.Dispose();
 
                 this.logger.Write(LogLevel.Verbose, "Named pipe server has been disposed.");
             }
+            if (this.outPipeServer != null)
+            {
+                this.logger.Write(LogLevel.Verbose, $"Named out pipe server {outPipeServer} shutting down...");
+
+                this.outPipeServer.Dispose();
+
+                this.logger.Write(LogLevel.Verbose, $"Named out pipe server {outPipeServer} has been disposed.");
+            }
         }
 
         private void ListenForConnection()
         {
-            Task.Factory.StartNew(
-                async () =>
+            var connectionTasks = new List<Task> {WaitForConnectionAsync(this.inOutPipeServer)};
+            if (this.outPipeServer != null)
+            {
+                connectionTasks.Add(WaitForConnectionAsync(this.outPipeServer));
+            }
+
+            Task.Run(async () =>
+            {
+                try
                 {
-                    try
-                    {
+                    await Task.WhenAll(connectionTasks);
+                    this.OnClientConnect(new NamedPipeServerChannel(this.inOutPipeServer, this.outPipeServer, this.logger));
+                }
+                catch (Exception e)
+                {
+                    this.logger.WriteException(
+                        "An unhandled exception occurred while listening for a named pipe client connection",
+                        e);
+
+                    throw e;
```

Comment:
Rethrow should never specify the exception - you'll lose stack info. Use `throw;` by itself.  Ideally our logger methods would return false so you could use C# exception filters e.g.:
```
catch (Exception e) when (this.logger.WriteException("...", e)) 
{
}
```
But we don't do that - yet.  @rjmholt Any chance we could modify the log methods to return `false` for this reason?

Reply:
Yeah, we could definitely modify the methods.

Or else write wrapper methods that explain the intended caller context:
```csharp
catch (Exception e) when (_logger.WriteExceptionWithoutUnwinding("...", e))
{
}
```

Bit of a mouthful -- just some name that says "I return false for use with exception filters".